cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/RecordWriter.java

```diff
@@ -110,20 +132,47 @@ public StreamResult write(Iterator<Tuple2<DecoratedKey, Object[]>> sourceIterato
         Map<String, Object> valueMap = new HashMap<>();
         try
         {
+            List<RingInstance> exclusions = failureHandler.getFailedInstances();
+            Set<Range<BigInteger>> newRanges = initialTokenRangeMapping.getRangeMap().asMapOfRanges().entrySet()
+                                                                       .stream()
+                                                                       .filter(e -> !exclusions.contains(e.getValue()))
+                                                                       .map(Map.Entry::getKey)
+                                                                       .collect(Collectors.toSet());
+
             while (dataIterator.hasNext())
             {
+                Tuple2<DecoratedKey, Object[]> rowData = dataIterator.next();
+                streamSession = maybeCreateStreamSession(taskContext, streamSession, rowData, newRanges, failureHandler);
+
+                sessions.add(streamSession);
```

Comment:
This was done to separate the session closures instead of having it scattered with checks for non-existent sessions. 

However, on looking further it seems like we do not really need to eagerly create the session for the partition's token range. Will update for it to be created lazily.